gateway/ln-gateway/src/gateway_module_v2/send_sm.rs

```diff
@@ -111,14 +122,39 @@ impl SendStateMachine {
     async fn send_payment(
         context: GatewayClientContextV2,
         max_delay: u64,
-        max_fee_msat: u64,
+        invoice_msats: u64,
         invoice: Bolt11Invoice,
-    ) -> Result<[u8; 32], String> {
+        contract_amount: Amount,
+    ) -> Result<[u8; 32], Cancelled> {
+        // The following three checks may fail in edge cases since they have inherent
+        // timing assumptions. Therefore, they may only be checked after we have created
+        // the state machine such that we can cancel the contract.
+        if invoice.is_expired() {
+            return Err(Cancelled::InvoiceExpired);
+        }
+
+        if max_delay == 0 {
+            return Err(Cancelled::TimeoutTooClose);
+        }
+
+        let min_contract_amount = context
+            .gateway
+            .payment_fees_v2()
+            .send
+            .add_fee(invoice_msats);
+
+        if contract_amount < min_contract_amount {
+            return Err(Cancelled::Underfunded);
+        }
+
+        let excess_fee = contract_amount - min_contract_amount;
+        let max_fee_msat = excess_fee.msats + (min_contract_amount.msats - invoice_msats) / 2;
```

Comment:
Why divide by 2 here? Is there some inherent assumption that half of the fee budget will be guaranteed profit? If so that's a bad approach imo: just specify the min profit and then use the remaining fee paid for routing if necessary.

Reply:
I am open to other fee strategies though.

Reply:
Just feels like a convoluted way of thinking about it, the gateway could just suggest a total fee and enforce a min fee that always goes to the gateway. Then you just subtract the min fee from the total fee and that's your routing budget.

Reply:
Well that is what we are doing here just that the total fee is twice the min fee and total fee is also enforced as a minimum. We then just let the client know the total fee instead of both / the exact split.

If we want to do a simplification here, I would suggest to get rid of overpayment that we enable now; instead we do it the same way as for receiving where there is one exact fee you have to pay and that's it. So both send and receive fee concepts function identical. Then the gateway is responsible for choosing a total fee that has the right tradeoff between reliability and fee for its users.

Reply:
https://github.com/fedimint/fedimint/pull/4975

Reply:
I fundamentally disagree with taking agency away from the user regarding the fee they are willing to pay. Some UIs may do this because they optimize for average users, but I still want to be able to build power-user tools that let you choose.

The fee suggestion by the gateway is a hack that will never be optimal, so at least we should retain some flexibility around it. Eventually I'd like to build an API where you can instruct the gateway to probe to more correctly guess the appropriate fee upfront.

Reply:
Could you please check the pr above for this, and repost it there if applicable. I kept the gateway flexible to allow for  additional fees but set the default client behaviour to no additional fees.